build/build.go

```diff
@@ -219,30 +274,109 @@ func Build(ctx context.Context, drivers []DriverInfo, opt map[string]Options, pw
 	return resp, nil
 }
 
-func LoadInputs(inp Inputs, target *client.SolveOpt) error {
+func createTempDockerfile(r io.Reader) (string, error) {
+	dir, err := ioutil.TempDir("", "dockerfile")
+	if err != nil {
+		return "", err
+	}
+	f, err := os.Create(filepath.Join(dir, "Dockerfile"))
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	if _, err := io.Copy(f, r); err != nil {
+		return "", err
+	}
+	return dir, err
+}
+
+func LoadInputs(inp Inputs, target *client.SolveOpt) (func(), error) {
 	if inp.ContextPath == "" {
-		return errors.New("please specify build context (e.g. \".\" for the current directory)")
+		return nil, errors.New("please specify build context (e.g. \".\" for the current directory)")
 	}
 
 	// TODO: handle stdin, symlinks, remote contexts, check files exist
 
-	if inp.DockerfilePath == "" {
-		inp.DockerfilePath = filepath.Join(inp.ContextPath, "Dockerfile")
+	var (
+		err              error
+		dockerfileReader io.Reader
+		dockerfileDir    string
+		dockerfileName   = inp.DockerfilePath
+		toRemove         []string
+	)
+
+	switch {
+	case inp.ContextPath == "-":
+		if inp.DockerfilePath == "-" {
+			return nil, errStdinConflict
+		}
+
+		buf := bufio.NewReader(os.Stdin)
+		magic, err := buf.Peek(archiveHeaderSize * 2)
+		if err != nil && err != io.EOF {
+			return nil, errors.Wrap(err, "failed to peek context header from STDIN")
+		}
+
+		if isArchive(magic) {
+			// stdin is context
+			up := uploadprovider.New()
+			target.FrontendAttrs["context"] = up.Add(buf)
+			target.Session = append(target.Session, up)
+		} else {
+			if inp.DockerfilePath != "" {
+				return nil, errDockerfileConflict
+			}
+			// stdin is dockerfile
+			dockerfileReader = buf
+			inp.ContextPath, _ = ioutil.TempDir("", "empty-dir")
+			toRemove = append(toRemove, inp.ContextPath)
+			target.LocalDirs["context"] = inp.ContextPath
+		}
+
+	case isLocalDir(inp.ContextPath):
+		target.LocalDirs["context"] = inp.ContextPath
+		switch inp.DockerfilePath {
+		case "-":
+			dockerfileReader = os.Stdin
+		case "":
+			dockerfileDir = inp.ContextPath
+		default:
+			dockerfileDir = filepath.Dir(inp.DockerfilePath)
```

Comment:
you shouldn't do this as it exposes a directory potentially outside of context. Also this does not work with symlinks.